integration/internal/container/exec.go

```diff
@@ -0,0 +1,90 @@
+package container
+
+import (
+	"bytes"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+	"golang.org/x/net/context"
+)
+
+type execResult struct {
+	exitCode  int
+	outBuffer *bytes.Buffer
+	errBuffer *bytes.Buffer
+}
+
+// Stdout returns stdout output of a command run by Exec()
+func (res *execResult) Stdout() string {
+	return res.outBuffer.String()
+}
+
+// Stderr returns stderr output of a command run by Exec()
+func (res *execResult) Stderr() string {
+	return res.errBuffer.String()
+}
+
+// Combined returns combined stdout and stderr output of a command run by Exec()
+func (res *execResult) Combined() string {
+	return res.outBuffer.String() + res.errBuffer.String()
+}
+
+// ExitCode returns exit code of a command run by Exec()
+func (res *execResult) ExitCode() int {
+	return res.exitCode
+}
+
+// Exec executes a command inside a container, returning the result
+// containing stdout, stderr, and exit code. Note:
+//  - this is a synchronous operation;
+//  - cmd stdin is closed.
+func Exec(ctx context.Context, cli client.APIClient, id string, cmd []string) (execResult, error) {
```

Comment:
I'm worried that this is really high level and will end up with other similar implementations to get at the lower level details.
I'm wondering if this can take an ExecConfig and return a hijack response (or some suitable wrapper)?

I think the main issue is that it wasn't clear that `ExecAttach` and `ExecStart` are really just calling the same endpoint and this is very different from the container API.

Reply:
The whole point was make it easy to use for the common case. A lower-level functions can definitely be added, too, in case there will be a need for those (so far I haven't seen it, but it might change).

> I think the main issue is that it wasn't clear that ExecAttach and ExecStart are really just calling the same endpoint

There were other issues, too, the one I remember is not closing something that should be closing.

When a test case is being developed, one should concentrate on the test case itself, not the intricacies of how to correctly execute a simple command inside a container and get the result.